SIMD: Add new universal intrinsic for ceil

howjmay opened the pull request:
Implement ceil SIMD operations. 
And NPYV_IMPL_AVX512_FROM_AVX2_PS_1ARG and
NPYV_IMPL_AVX512_FROM_AVX2_PD_1ARG are added for constructing not existing 512-bits intrinsics. 
The error in npyv512_combine_ps256 has been fixed too.

HowJMay commented:
Sorry @seiko2plus I have some problem reading the error message in [CI](https://github.com/numpy/numpy/pull/19022/checks?check_run_id=2704066507)

When I saw `Test_SIMD_FP_128_baseline_f32`, I am not sure which architecture it is using. Is it `SSE`?

The other thing is there is no direct way to truncate a double precision floating number, so I choose to use a pure c implementation. Is it ok?

seiko2plus reviewed (commented):
Sorry for the delayed response, the baseline features are configured through build option `--cpu-baseline`, the default setting on x86_64 is equivalent to `SSE SSE2 SSE3`.

The current CI error is related to the emulated version of `npyv_ceil_f32`, see my comment blow for more clarification.

howjmay committed:
ENH: Add conversion for 512bit intrin to 256bit intrin

512-bits intrinsics sometimes not exist in some operations. We can
construct the 512-bits intrinsics with two 256-bits intrinsics.
Therfore, NPYV_IMPL_AVX512_FROM_AVX2_PS_1ARG and
NPYV_IMPL_AVX512_FROM_AVX2_PD_1ARG are added. And the error in
npyv512_combine_ps256 has been fixed too.

howjmay committed:
ENH: Implement SIMD for ceil

HowJMay commented on numpy/core/src/common/simd/neon/math.h:
```diff
@@ -153,4 +153,10 @@ NPY_FINLINE npyv_s64 npyv_min_s64(npyv_s64 a, npyv_s64 b)
 
+// ceil
+#if NPY_SIMD_F64
+    #define npyv_ceil_f32 vrndpq_f32
```
thanks!

HowJMay commented on numpy/core/tests/test_simd.py:
```diff
@@ -328,7 +328,22 @@ def test_square(self):
+
+        data_ceil = [math.ceil(x) for x in data]
+        ceil = self.ceil(vdata)
+        assert ceil == data_ceil
```
thanks!

HowJMay commented on numpy/core/src/common/simd/sse/math.h:
```diff
@@ -143,4 +143,24 @@ NPY_FINLINE npyv_s64 npyv_min_s64(npyv_s64 a, npyv_s64 b)
+    NPY_FINLINE npyv_f64 npyv_ceil_f64(npyv_f64 a)
+    {
+        double *_a = (double *) &a;
+        return _mm_set_pd(ceil(_a[1]), ceil(_a[0]));
```
thank you for this information!

HowJMay commented on numpy/core/src/common/simd/sse/math.h:
```diff
@@ -143,4 +143,24 @@ NPY_FINLINE npyv_s64 npyv_min_s64(npyv_s64 a, npyv_s64 b)
+        npyv_f32 has_trunc = _mm_cmplt_ps(trunc_a, a);
+        npyv_f32 ceil_res = _mm_add_ps(trunc_a, _mm_and_ps(has_trunc, _mm_set_ps(1.0, 1.0, 1.0, 1.0)));
+        return npyv_select_f32(_mm_castps_si128(a_not_nan), a, ceil_res);
+    }
```
thank you for explaining

seiko2plus commented on numpy/core/src/common/simd/neon/math.h:
```diff
@@ -153,4 +153,20 @@ NPY_FINLINE npyv_s64 npyv_min_s64(npyv_s64 a, npyv_s64 b)
+#ifdef NPY_HAVE_ASIMD
+    #define npyv_ceil_f32 vrndpq_f32
+#else
+   NPY_FINLINE npyv_f32 npyv_ceil_f32(npyv_f32 a)
+   {
+       npyv_f32 conv_trunc = vcvtq_f32_s32(vcvtq_s32_f32(a));
+       npyv_f32 conv_trunc_add_one = npyv_add_f32(conv_trunc, vdupq_n_f32(1.0f));
+       npyv_u32 mask = vcltq_f32(conv_trunc, a);
+       return vbslq_f32(mask, conv_trunc, conv_trunc_add_one);
+   }
```
```suggestion
NPY_FINLINE npyv_f32 npyv_ceil_f32(npyv_f32 a)
    {
        const npyv_s32 szero = vreinterpretq_s32_f32(vdupq_n_f32(-0.0f));
        const npyv_u32 one = vreinterpretq_u32_f32(vdupq_n_f32(1.0f));
        const npyv_s32 max_int = vdupq_n_s32(0x7fffffff);
        /**
         * On armv7, vcvtq.f32 handles special cases as follows:
         *  NaN return 0
         * +inf or +outrange return 0x80000000(-0.0f)
         * -inf or -outrange return 0x7fffffff(nan)
         */
        npyv_s32 roundi = vcvtq_s32_f32(a);
        npyv_f32 round = vcvtq_f32_s32(roundi);
        npyv_f32 ceil = vaddq_f32(round, vreinterpretq_f32_u32(
            vandq_u32(vcltq_f32(round, a), one))
        ); 
        // respect signed zero, e.g. -0.5 -> -0.0
        npyv_f32 rzero = vreinterpretq_f32_s32(vorrq_s32(
            vreinterpretq_s32_f32(ceil),
            vandq_s32(vreinterpretq_s32_f32(a), szero)
        ));
        // if nan or overflow return a
        npyv_u32 nnan = npyv_notnan_f32(a);
        npyv_u32 overflow = vorrq_u32(
            vceqq_s32(roundi, szero), vceqq_s32(roundi, max_int)
        );
        return vbslq_f32(vbicq_u32(nnan, overflow), rzero, a);
    }

```

Similar to the other architectures, ceil_f32 should respect special cases and signed zero.
My suggestion here is similar to the SSE version except for handling finite and outranges due to the nature of `vcvtq_s32_f32`.

Note that CI doesn't cover armv7/neon neither aarch32/ASIMD so I had to test it on my local machine and it seems good to me.

seiko2plus replied:
update: we have now a CI/workflow for testing universal intrinsics on armv7/aarch32.

seiko2plus reviewed (changes requested):
Just one case left, the ceil implementation of Armv7/Neon should respect the special cases, and signed zero similar to other extensions.

seiko2plus commented on numpy/core/src/common/simd/avx512/math.h:
```diff
@@ -112,4 +112,8 @@ NPY_FINLINE npyv_f64 npyv_minp_f64(npyv_f64 a, npyv_f64 b)
 
+// ceil
+NPYV_IMPL_AVX512_FROM_AVX2_PS_1ARG(npyv_ceil_f32, _mm256_ceil_ps)
+NPYV_IMPL_AVX512_FROM_AVX2_PD_1ARG(npyv_ceil_f64, _mm256_ceil_pd)
```
no problem, let's keep them then. thank you

seiko2plus committed:
SIMD: Fix impl of intrinsic `npyv_ceil_f32` on armv7/neon

seiko2plus reviewed (approved):
Thank you Yang Hau. We still need to add intrinsics for `floor, trunc, rint` so we can get ride of the following raw x86 SIMD.

https://github.com/numpy/numpy/blob/fa397e9d9a821a178ad620b28bb334b800d2c93d/numpy/core/src/umath/simd.inc.src#L1400-L1411

HowJMay commented:
> Thank you Yang Hau. We still need to add intrinsics for `floor, trunc, rint` so we can get ride of the following raw x86 SIMD.
> 
> https://github.com/numpy/numpy/blob/fa397e9d9a821a178ad620b28bb334b800d2c93d/numpy/core/src/umath/simd.inc.src#L1400-L1411

Let me work on them

HowJMay commented:
I will work on the following intrinsics once this PR being merge!

HowJMay commented:
Hi @seiko2plus, sorry for disturbing. I just found some inconsistency in the implementation of NEON's one and SSE's one. When encountering `NAN`, `±INF`, I think the correct behavior is returning the input argument `a`? If so, I should add the protection on other architectures. One thing I am curious is, on some architectures, we simply implement `ceil` with macro, then, should we add protection to these functions?

seiko2plus commented:
> I will work on the following intrinsics once this PR being merge!

let's wait +1 day more before merging this pr, to give chance to other contributors to review it.

> When encountering NAN, ±INF, I think the correct behavior is returning the input argument a?

I'm not sure exactly what you mean, but this patch already handles the following for all extensions:
- If arg is ±∞, it is returned unmodified
- If arg is ±0, it is returned, unmodified
- If arg is NaN, NaN is returned

In the case of `FE_INEXACT` which I think is what you mean. Yes, we return the arg(`a`) for the emulated intrinsics of `NEON` and `SSE<4.1`when encountering `overflow` without raising `FE_INEXACT exception` which can be acceptable.

>  If so, I should add the protection on other architectures.

no, non-emulated intrinsics(native) support IEEE floating-point arithmetic (IEC 60559), and correctly handle `FE_INEXACT`.

seiko2plus commented:
@mattip, ping

HowJMay commented:
Hi @seiko2plus 
What I worry is we are not handling `NAN` case under `SSE<4.1`. Maybe I miss something?

seiko2plus commented:
Hi @HowJMay, Yes, it handles both nan/inf cases please check my comment https://github.com/numpy/numpy/pull/19022#discussion_r685587127

seiko2plus commented:
> What I worry is we are not handling NAN case under SSE<4.1.

Why being worry? we already have test cases for it.

HowJMay commented:
> > What I worry is we are not handling NAN case under SSE<4.1.
> 
> Why being worry? we already have test cases for it.

That's true haha 😆😆😆😆

mattip commented:
Is there a benchmark run that shows the improvement? How much does this inflate the `_multiarray_umath` shared object?

seiko2plus commented:
@mattip, This pr only add universal intrinsics for ceil, so it can be used to replace the current existence raw SIMD in:

https://github.com/numpy/numpy/blob/fa397e9d9a821a178ad620b28bb334b800d2c93d/numpy/core/src/umath/simd.inc.src#L1400-L1411

mattip commented:
Thanks @HowJMay and for the review @seiko2plus
